cmd/cosign/cli/verify_attestation.go

```diff
@@ -166,8 +191,65 @@ func (c *VerifyAttestationCommand) Exec(ctx context.Context, args []string) (err
 		co.VerifyBundle = false
 
 		verified, err := cosign.Verify(ctx, ref, co)
-		if err != nil {
-			return err
+
+		for _, vp := range verified {
+			var payloadData map[string]interface{}
+			err := json.Unmarshal(vp.Payload, &payloadData)
+			if err != nil {
+				return err
+			}
+
+			if predicateTypeMap[c.PredicateType] != payloadData["payloadType"] {
+				continue
+			}
+
+			decodedPayload, err := base64.StdEncoding.DecodeString(payloadData["payload"].(string))
+
+			if err != nil {
+				return err
+			}
+
+			switch c.PredicateType {
+			case predicateCustom:
+				var cosignStatement in_toto.Statement
+				if err := json.Unmarshal(decodedPayload, &cosignStatement); err != nil {
+					return err
+				}
+				payload, _ := json.Marshal(cosignStatement.Predicate)
+				if err := cue.ValidateJSON(payload, c.Policies.entrypoints); err != nil {
+					return err
+				}
+			case predicateLink:
+				var linkStatement in_toto.LinkStatement
+				if err := json.Unmarshal(decodedPayload, &linkStatement); err != nil {
+					return err
+				}
+				payload, _ := json.Marshal(linkStatement.Predicate)
+				if err := cue.ValidateJSON(payload, c.Policies.entrypoints); err != nil {
+					return err
+				}
+			case predicateSlsa:
+				var slsaProvenanceStatement in_toto.ProvenanceStatement
+				if err := json.Unmarshal(decodedPayload, &slsaProvenanceStatement); err != nil {
+					return err
+				}
+				payload, _ := json.Marshal(slsaProvenanceStatement.Predicate)
+				if err := cue.ValidateJSON(payload, c.Policies.entrypoints); err != nil {
+					return err
+				}
+			case predicateSpdx:
+				var spdxStatement in_toto.SPDXStatement
+				if err := json.Unmarshal(decodedPayload, &spdxStatement); err != nil {
+					return err
+				}
+				payload, _ := json.Marshal(spdxStatement.Predicate)
+				if err := cue.ValidateJSON(payload, c.Policies.entrypoints); err != nil {
+					return err
+				}
+			default:
+				continue
```

Comment:
Is `c.PredicateType` already validated at this point? Or should an error be thrown about an unknown predicate type? (for example when a user has a typo)

If not validated, this looks to silently pass the check.

Reply:
thank you so much @verdverm, it should be fixed right now.